test/upgrade.js

```diff
@@ -791,7 +799,587 @@ test('autobase upgrade - consensus 3 writers', async t => {
   await confirm([a1, b1])
 
   t.is(b1.view.signedLength, 5) // majority can continue
-  t.is((await b1.system.getIndexedInfo()).version, b1.version)
+  t.is((await b1.system.getIndexedInfo()).version, version + 1)
+})
+
+test('autobase upgrade - downgrade', async t => {
+  const [s1] = await createStores(1, t)
+
+  const a0 = new Autobase(s1.session(), null, {
+    apply,
+    open: store => store.get('test', { valueEncoding: 'json' }),
+    valueEncoding: 'json'
+  })
+
+  const version = a0.maxSupportedVersion
+
+  await a0.ready()
+
+  await a0.append({ data: 'version 0' })
+
+  t.is(a0.view.indexedLength, 1)
+
+  await a0.close()
+
+  const a1 = new Autobase(s1.session(), a0.bootstrap, {
+    apply,
+    open: store => store.get('test', { valueEncoding: 'json' }),
+    valueEncoding: 'json'
+  })
+
+  a1.maxSupportedVersion = version + 1
+
+  await a1.ready()
+
+  t.is(a1.view.indexedLength, 1)
+
+  await a1.append({ data: 'version 1' })
+
+  t.is(a1.view.indexedLength, 2)
+
+  t.is(a1.version, version + 1)
+  t.is((await a1.system.getIndexedInfo()).version, version + 1)
+
+  await a1.close()
+
+  t.not(await a1.local.getUserData('autobase/system'), null)
+
+  // go back to previous version
+  const fail = new Autobase(s1.session(), a0.bootstrap, {
+    apply,
+    open: store => store.get('test', { valueEncoding: 'json' }),
+    valueEncoding: 'json'
+  })
+
+  await t.exception(fail.ready())
+
+  t.is(await fail.local.getUserData('autobase/system'), null)
+})
+
+test('autobase upgrade - downgrade then restart', async t => {
+  const [s1] = await createStores(1, t)
+
+  const a0 = new Autobase(s1.session(), null, {
+    apply,
+    open: store => store.get('test', { valueEncoding: 'json' }),
+    valueEncoding: 'json'
+  })
+
+  const version = a0.maxSupportedVersion
+
+  await a0.ready()
+
+  await a0.append({ data: 'version 0' })
+
+  t.is(a0.view.indexedLength, 1)
+
+  await a0.close()
+
+  const a1 = new Autobase(s1.session(), a0.bootstrap, {
+    apply,
+    open: store => store.get('test', { valueEncoding: 'json' }),
+    valueEncoding: 'json'
+  })
+
+  a1.maxSupportedVersion = version + 1
+
+  await a1.ready()
+
+  t.is(a1.view.indexedLength, 1)
+
+  await a1.append({ data: 'version 1' })
+
+  t.is(a1.view.indexedLength, 2)
+
+  t.is((await a1.system.getIndexedInfo()).version, version + 1)
+
+  await a1.close()
+
+  t.not(await a1.local.getUserData('autobase/system'), null)
+
+  // go back to previous version
+  const fail = new Autobase(s1.session(), a0.bootstrap, {
+    apply,
+    open: store => store.get('test', { valueEncoding: 'json' }),
+    valueEncoding: 'json'
+  })
+
+  await t.exception(fail.ready())
+
+  t.is(await fail.local.getUserData('autobase/system'), null)
+
+  // go back to previous version
```

Comment:
Note: `fail` was never closed. I actually don't think it can be closed (because `close` would wait on `ready`), but this does mean that `fail`'s session never got closed (not 100% sure if that's ok)